clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -809,15 +831,25 @@ public boolean isDone() {
         private final Callback userCallback;
         private final ProducerInterceptors<K, V> interceptors;
         private final TopicPartition tp;
+        private final TransactionState transactionState;
 
         public InterceptorCallback(Callback userCallback, ProducerInterceptors<K, V> interceptors,
-                                   TopicPartition tp) {
+                                   TopicPartition tp, TransactionState transactionState) {
             this.userCallback = userCallback;
             this.interceptors = interceptors;
             this.tp = tp;
+            this.transactionState = transactionState;
         }
 
         public void onCompletion(RecordMetadata metadata, Exception exception) {
+            if (exception != null && exception instanceof OutOfOrderSequenceException) {
+                // This is a fatal exception. All future send / sendOffsets / beginTransaction, endTransaction, etc.
+                // should be aborted with an IllegalStateException.
+                if (transactionState == null) {
+                    throw new IllegalStateException("Received OutOfOrderSequenceException when local state for sequence numbers doesn't exist.");
```

Comment:
I'd suggest throwing this exception still as an `OutOfOrderSequenceException` or sth. that could be distinguished with other `IllegalStateException`. The reason is that callers, like Streams, need to special handle this case compared with other possibly thrown exceptions. If it is thrown to users as an `IllegalStateException` then it is a bit hard to do and we have to rely on its error message.